qiskit_experiments/curve_analysis/fit_models.py

```diff
@@ -0,0 +1,257 @@
+        return f"{self.__class__.__name__}(x, {sigrepr})"
+
+
+class SingleFitFunction(FitModel):
+    r"""Fit model consisting of a single curve.
+
+    This model is created when only single curve exist in the fit model.
+
+    .. math::
+
+        F(x, \Theta) = f(x, \vec{p})
+
+    The parameter :math:`\vec{p} = \Theta \cup \Theta_{\rm fix}` which is a union of
+    the fit parameters and the fixed parameters :math:`\Theta_{\rm fix}`.
+    The function :math:`f` is usually set by :attr:`SeriesDef.fit_func` which is
+    a standard python function.
+
+    .. seealso::
+
+        Class :class:`FitModel`.
+    """
+
+    def __call__(self, x: np.ndarray, *params) -> np.ndarray:
+        kwparams = dict(zip(self._union_params, params))
+        kwparams.update(self._fixed_params)
+
+        return self._fit_functions[0](x, **{p: kwparams[p] for p in self._signatures[0]})
+
+
+class CompositeFitFunction(FitModel):
+    r"""Fit model consisting of multiple curves sharing fit parameters.
+
+    This model is created when multiple curves form a fit model.
+
+    .. math::
+
+        F(x, \Theta) = f_0(x_0, \vec{p}_0) \oplus f_1(x_1, \vec{p}_1) \oplus ...
+
+    The parameter :math:`\vec{p}_i = \theta_i \cup \Theta_{\rm fix}` is a union of the
+    fit parameter for the function :math:`f_i` and the fixed parameters :math:`\Theta_{\rm fix}`.
+    The composite function :math:`F` consists of multiple fit functions :math:`f_i`
+    taking independent data points :math:`x_i` with partly shared fit parameters :math:`\theta_i`,
+    where :math:`\Theta = \theta_0 \cup \theta_1 \cup ...` and the composite data vector
+    :math:`x = x_0 \oplus x_1 \oplus ...`
+
+    In the NumPy array data, this can be represented
+    by a single array together with the array specifying location, which is provided as a
+    :attr:`CompositeFitFunction.data_allocation`. For example:
+
+    .. parsed-literal::
+
+        data_allocation = array([0, 0, 0, 0, 0, 1, 1, 1, 1, 1, ...])
+
+    This data represents the location where the function with index ``i``
+    is returned and where the x values :math:`x_i` comes from.
+    One must set this data indices before calling the composite fit function.
+
+    .. seealso::
+
+        Class :class:`FitModel`.
+    """
+
+    def __init__(
+        self,
+        fit_functions: List[Callable],
+        signatures: List[List[str]],
+        fit_models: Optional[List[str]] = None,
+        fixed_parameters: Optional[List[str]] = None,
+    ):
+        super().__init__(fit_functions, signatures, fit_models, fixed_parameters)
+        self.data_allocation = None
+
+    def __call__(self, x: np.ndarray, *params) -> np.ndarray:
+        kwparams = dict(zip(self._union_params, params))
+        kwparams.update(self._fixed_params)
+
+        y = np.zeros(x.size, dtype=float)
+        for i, (func, sig) in enumerate(zip(self._fit_functions, self._signatures)):
+            inds = self.data_allocation == i
```

Comment:
Maybe verify that `self.data_allocation` is not `None` anymore?

Reply:
This is called multiple times from scipy fitter. Adding redundant branching logic may degrade performance (I didn't check though)

Reply:
Note that usually this model is not called by end-users or custom curve analysis. But the logic calling the model exist in the base class of curve analysis. This means too much redundancy for readability or edge case will just degrade performance.

Reply:
Note that this is boolean index array. `data_allocation` is not a dict.
```
>>> vec = np.array([0, 0, 0, 0, 1, 0, 1, 1, 1])
>>> vec[1] 
0
>>> vec == 1
array([False, False, False, False,  True, False,  True,  True,  True])
```